File: bindings/python/samples/stream_recv_sample.py
import socket

import zlink
from sample_support import tcp_endpoint, wait_socket_monitor_event


def main():
# --8<-- [start:doc]
    port, endpoint = tcp_endpoint()

    with zlink.create_context() as ctx:
        with zlink.create_stream_socket(ctx) as server:
            with server.monitor_open(zlink.MonitorEventMask.ACCEPTED) as server_monitor:
                server.bind(endpoint)
                with socket.create_connection(("127.0.0.1", port), timeout=3.0) as client:
                    wait_socket_monitor_event(server_monitor, zlink.MonitorEventMask.ACCEPTED)
                    client.sendall(b"hello-stream")
                    received = zlink.create_received()
                    if not server.recv_into(received):
                        raise AssertionError("expected stream payload")
                    with received:
                        if received.to_bytes_list() != [b"hello-stream"]:
                            raise AssertionError("unexpected stream payload")
                        if not received.routing_id:
                            raise AssertionError("stream sample expected a routing id")
                        server.send(received.routing_id).message(b"hello-stream").submit()
                    reply = client.recv(64)
                    if reply != b"hello-stream":
                        raise AssertionError(f"unexpected stream reply: {reply!r}")
            print('[stream/recv] send: "hello-stream" → recv: "hello-stream"')
# --8<-- [end:doc]


if __name__ == "__main__":
    main()

File: bindings/rust/samples/stream_recv_sample.rs
//! STREAM direct recv sample – demonstrates STREAM socket with direct recv.
//! The STREAM socket binds as a server; a raw TCP client connects inward.

mod sample_support;

use std::io::{Read, Write};
use std::net::TcpStream;

use zlink::{Context, SocketMonitor};

fn main() {
    // --8<-- [start:doc]
    let ctx = Context::new().expect("context creation failed");

    let stream = ctx.stream_socket().expect("stream socket failed");
    stream.bind("tcp://127.0.0.1:0").expect("bind failed");
    let endpoint = stream.last_endpoint().expect("last_endpoint failed");
    let stream_mon = SocketMonitor::open(&stream).expect("stream monitor open failed");

    let tcp_addr = endpoint.strip_prefix("tcp://").unwrap();
    let mut tcp_client = TcpStream::connect(tcp_addr).expect("tcp connect failed");
    tcp_client.set_nodelay(true).expect("set_nodelay failed");
    sample_support::wait_stream_connected(&stream_mon);
    drop(stream_mon);

    tcp_client
        .write_all(b"hello-stream")
        .expect("tcp write failed");
    tcp_client.flush().expect("tcp flush failed");

    let mut received = zlink::Received::empty();
    stream
        .recv(&mut received, zlink::RecvFlags::NONE)
        .expect("server recv failed");
    assert_eq!(received.parts()[0].as_bytes(), b"hello-stream");
    received
        .send()
        .message(zlink::Message::try_from(b"hello-stream").expect("reply message failed"))
        .submit()
        .expect("stream reply failed");
    let mut response = [0u8; 12];
    tcp_client
        .read_exact(&mut response)
        .expect("tcp read failed");
    assert_eq!(&response, b"hello-stream");
    println!(
        "[stream/recv] send: \"hello-stream\" → recv: \"{}\"",
        received.parts()[0].as_str().unwrap()
    );
    // --8<-- [end:doc]
}
